### PR TITLE
Fix '-PipelineVariable' to set variable in the right scope

### LINKED ISSUE #16155: PipelineVariable throws error if variable name used exists in function definition
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

Sample function:

```powershell
function Invoke-AddOne {
    param (
        [Parameter(ValueFromPipeline)]
        [string[]]$Number
    )

    Process {
        $testValue = $_ + 1
        $testValue
    }
}
```

Run each command to recreate error:

```powershell
1,2,3 | Invoke-AddOne -PV test # does not throw error
1,2,3 | Invoke-AddOne -PV testValue # does throw error
```

If this is a matter of me writing an improper example, please let me know. I see the error mentions the use of `Remove-Variable`, but I've never heard of any recommendations to use `Remove-Variable` for custom written functions. If this the case and I just need to add `Remove-Variable` to my functions, then my apologies.

### Expected behavior

```console
When the function is called while supplying a `-PipelineVariable` value, it should work as normal, storing the pipeline value within the variable name to be used downstream in the pipeline.


1,2,3 | Invoke-AddOne -PV test | % { $test }
2
3
4


Does not return any errors
```

### Actual behavior

```console
When a variable name that is used in the definition of the function is supplied as a `-PipelineVariable` variable name, it throws an error.


1,2,3 | Invoke-AddOne -PV testValue | % { $testValue }
Invoke-AddOne: Cannot remove variable testValue because the variable has been optimized and is not removable. Try using the Remove-Variable cmdlet (without any aliases), or dot-sourcing the command that you are using to remove the variable.
```

### Error details

```console
Exception             :
    Type        : System.Management.Automation.SessionStateUnauthorizedAccessException
    ErrorRecord :
        Exception             :
            Type    : System.Management.Automation.ParentContainsErrorRecordException
            Message : Cannot remove variable testValue because the variable has been optimized and is not removable. Try using the Remove-Variable cmdlet (without any aliases), or dot-sourcing the command that you are using to remove the variable.
            HResult : -2146233087
        TargetObject          : testValue
        CategoryInfo          : WriteError: (testValue:String) [], ParentContainsErrorRecordException
        FullyQualifiedErrorId : VariableNotRemovableRare
    ItemName    : testValue
    TargetSite  :
        Name          : RemoveVariable
        DeclaringType : System.Management.Automation.SessionStateScope, System.Management.Automation, Version=7.2.0.9, Culture=neutral, PublicKeyToken=31bf3856ad364e35
        MemberType    : Method
        Module        : System.Management.Automation.dll
    Message     : Cannot remove variable testValue because the variable has been optimized and is not removable. Try using the Remove-Variable cmdlet (without any aliases), or dot-sourcing the command that you are using to remove the variable.
    Source      : System.Management.Automation
    HResult     : -2146233087
    StackTrace  :
   at System.Management.Automation.SessionStateScope.RemoveVariable(String name, Boolean force) in System.Management.Automation.dll:token 0x6003bdf+0x46
   at System.Management.Automation.SessionStateInternal.RemoveVariable(String name, Boolean force) in System.Management.Automation.dll:token 0x6003b9d+0x2a
   at System.Management.Automation.PSScriptCmdlet.RunClause(Action`1 clause, Object dollarUnderbar, Object inputToProcess) in System.Management.Automation.dll:token 0x600368b+0x125
   at System.Management.Automation.PSScriptCmdlet.DoProcessRecord() in System.Management.Automation.dll:token 0x6003687+0x3b
   at System.Management.Automation.CommandProcessor.ProcessRecord() in System.Management.Automation.dll:token 0x6002100+0x1ae
TargetObject          : testValue
CategoryInfo          : WriteError: (testValue:String) [Invoke-AddOne], SessionStateUnauthorizedAccessException
FullyQualifiedErrorId : VariableNotRemovableRare,Invoke-AddOne
InvocationInfo        :
    MyCommand        : Invoke-AddOne
    ScriptLineNumber : 1
    OffsetInLine     : 9
    HistoryId        : 10
    Line             : 1,2,3 | Invoke-AddOne -PV testValue | % { $testValue }
    PositionMessage  : At line:1 char:9
                       + 1,2,3 | Invoke-AddOne -PV testValue | % { $testValue }
                       +         ~~~~~~~~~~~~~~~~~~~~~~~~~~~
    InvocationName   : Invoke-AddOne
    CommandOrigin    : Internal
ScriptStackTrace      : at <ScriptBlock>, <No file>: line 1
```
```


### Environment data

```powershell
PS> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      7.2.0-preview.9
PSEdition                      Core
GitCommitId                    7.2.0-preview.9
OS                             Microsoft Windows 10.0.22000
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```
```


### Visuals

_No response_

### PR DESCRIPTION
**Pull request by daxian-dbw**
# PR Summary

Fix #16155

Fix '-PipelineVariable' to set variable in the right scope and to not remove the pipeline variable at the end of pipeline when it's actually using an already existing variable.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.

### CODE REVIEW

File: src/System.Management.Automation/engine/MshCommandRuntime.cs [hunk 1]
```diff
@@ -983,19 +984,32 @@ internal void SetupPipelineVariable()
...
             // Create the pipeline variable
-            _pipelineVarReference = new PSVariable(this.PipelineVariable);
+            _pipelineVarReference = new PSVariable(PipelineVariable);
             _state.PSVariable.Set(_pipelineVarReference);
 
-            // Get the reference again in case we re-used one from the
-            // same scope.
-            _pipelineVarReference = _state.PSVariable.Get(this.PipelineVariable);
+            // Get the reference again in case that we need to re-use an existing one from the same scope.
+            // In the case that an existing variable from the same scope is used, don't attemtp to remove
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/MshCommandRuntime.cs:991]:
```suggestion
            // In the case that an existing variable from the same scope is used, don't attempt to remove
```

File: src/System.Management.Automation/engine/MshCommandRuntime.cs [hunk 2]
```diff
@@ -983,19 +984,32 @@ internal void SetupPipelineVariable()
...
             {
                 this.OutputPipe.SetPipelineVariable(_pipelineVarReference);
             }
         }
 
+        internal void RemovePipelineVariable()
+        {
+            if (_shouldRemovePipelineVariable)
+            {
+                // Remove pipeline varaiable when a pipeline is being torn down.
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/MshCommandRuntime.cs:1008]:
```suggestion
                // Remove pipeline variable when a pipeline is being torn down.
```

**Review by iSazonov — COMMENTED** [review]
Have we the same issue for other variables like OutVariable?

**Comment by daxian-dbw** [comment]
No, only pipeline variable is special.

File: src/System.Management.Automation/engine/pipeline.cs [hunk 3]
```diff
@@ -1298,7 +1298,16 @@ private void DisposeCommands()
                         // pipeline failure and continue disposing cmdlets.
                         try
                         {
-                            commandProcessor.CommandRuntime.RemoveVariableListsInPipe();
+                            var commandRuntime = commandProcessor.CommandRuntime;
```
Comment (rjmholt) [inline src/System.Management.Automation/engine/pipeline.cs:1301]: Please change from `var` here

File: src/System.Management.Automation/engine/pipeline.cs [hunk 4]
```diff
@@ -1298,7 +1298,16 @@ private void DisposeCommands()
...
+                            if (commandProcessor is CommandProcessor && commandProcessor.Command is not PSScriptCmdlet)
+                            {
+                                // Only a cmdlet can have variable lists defined via the common parameters.
+                                // We only need to take care of binary cmdlet here, because for script cmdlet,
+                                // the variable lists were already removed when exiting a scope.
+                                commandRuntime.RemoveVariableListsInPipe();
+                            }
```
Comment (rjmholt) [inline src/System.Management.Automation/engine/pipeline.cs:1308]: Is saving the variable remove operation going to be worth the reflection here? Just want to ask the question to confirm

File: src/System.Management.Automation/engine/MshCommandRuntime.cs [hunk 5]
```diff
@@ -983,19 +984,32 @@ internal void SetupPipelineVariable()
                 _state = new SessionState(Context.EngineSessionState);
 
             // Create the pipeline variable
-            _pipelineVarReference = new PSVariable(this.PipelineVariable);
+            _pipelineVarReference = new PSVariable(PipelineVariable);
             _state.PSVariable.Set(_pipelineVarReference);
```
[the remark this answers is not in the document]
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/MshCommandRuntime.cs:988]: Good point. Will update the code.

**Review by rjmholt — APPROVED** [review]

**Pull request merged**